scripts/index_tfidf

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import glob
+import os
+import pickle
+
+import numpy as np
+
+from sklearn.feature_extraction.text import CountVectorizer, HashingVectorizer, TfidfTransformer
+
+from aidistillery.file_handling import identifier_from_path
+
+
+def main(args):
+    """
+    Arguments
+    ---------
+    :args: Namespace object
+
+
+    """
+    raw_documents = glob.glob(os.path.join(args.datadir, '*'))
+
+    vect_args = {
+        'input': 'filename',
+        'stop_words': 'english',
+    }
+
+    if args.hash:
+        if args.n_features is None:
+            args.n_features = 2 ** 20  # sklearn default for hashingvectorizer
+        else:
+            print("Limiting vocab with hashing vectorizer cause hashes to collide")
+
+        vect = HashingVectorizer(n_features=args.n_features, norm=None, **vect_args)
+    else:
+        vect = CountVectorizer(max_features=args.n_features, **vect_args)                               
+
+    tfidf = TfidfTransformer(norm='l2')
+
+    X = vect.fit_transform(raw_documents)
+
+    X = tfidf.fit_transform(X)
+    ids = [identifier_from_path(d) for d in raw_documents]
+
+    os.makedirs(args.outpath, exist_ok=True)
+    with open(os.path.join(args.outpath, "vectorizer.pkl"), 'wb') as fhandle:
+        pickle.dump(vect, fhandle)
+
+    with open(os.path.join(args.outpath, "tfidf_sparse.pkl"), 'wb') as fhandle:
+        pickle.dump(X, fhandle)
+
+    with open(os.path.join(args.outpath, "index2identifier.pkl"), 'wb') as fhandle:
+        pickle.dump(ids, fhandle)
+
+
+
+
+
+
+
+def _add_args(parser):
+    parser.add_argument("-d", "--datadir", required=True,
+                        help="Path to txt file root")
+    parser.add_argument("-o", "--outpath", required=True,
+                        help="Path to output directory")
+    parser.add_argument("-n", "--n-features", default=None, type=int,
+                        help="Use that many features (default unlimited)")
+    parser.add_argument("--hash", action='store_true', default=False,
+                        help="Use (stateless) hashing vectorizer")
+
+if __name__ == "__main__":
+    import argparse
```

Comment:
Is there any reason you import stuff stuff other than at the top of the file? 
I kinda like the look and feel though

Reply:
At rare cases, it makes sense to do this. For this instance, argparse is only needed when the script is executed itself. When another part would import some function of it, argparse does not need to be imported. Other scenarios would be: you use matplotlib in one plot function but nowhere else in the module; then you do not know if the client will use the plotting function and you avoid one dependency as long as it is not really needed.